### PR TITLE
Normalize geometry collection geo shapes provided as object literals

### PR DESCRIPTION
**Pull request by BaurzhanSakhariev**
Transform geometry collection to `MultiPoint` or `MultiPoligon` if possible. This aligns it with WKT string behavior

Relates to https://github.com/crate/crate/pull/16761#discussion_r1806034489

### CODE REVIEW

File: server/src/main/java/io/crate/geo/GeoJSONUtils.java [hunk 1]
```diff
@@ -117,10 +117,11 @@ public static Map<String, Object> shape2Map(Shape shape) {
             for (Shape collShape : shapeCollection) {
                 geometries.add(shape2Map(collShape));
             }
-            return Map.of(
+            // Transform Geometry Collection back to MultiPoint/MultiPolygon if possible.
+            return GeoJSONUtils.validateGeoJson(Map.of(
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/geo/GeoJSONUtils.java:121]: Integration test (I will replace it later with unit tests) passes without it, ie comparison of literals, insert + select works.

`test_check_equality_with_itself`  fails without this change but not sure if test code path is really relevant to any user facing scenario:
compare object literal with normalized version which is `type.implicitCast(GeoJSONUtils.map2Shape(geoJson));`

I will look for a scenario where we can hit such combination.

**Comment by BaurzhanSakhariev** [comment]
Problem: collection_of_points != multi_point for object literals.
Another approach to make equality ^ work is to adjust another part.

Instead of geom_collection -> multi* make it multi* -> geom_collection. This way we will comply with the existing code and would need to tweak WKT string handlign instead.

It might be a bit tricky as we it happens in the third party library. 

Transformation geom_collection -> Multi* happens in `GeoPointType.WKT_READER.parse(wkt)`

File: server/src/main/java/io/crate/geo/GeoJSONUtils.java (hunk 1 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/geo/GeoJSONUtils.java]: >I will look for a scenario where we can hit such combination 

I was not able to find a user facing scenario where we need this change. 
This change was targeting a failing test originally suggested by @mfussenegger but I think we cannot call `shape2Map` with non-normalized values - the only user facing usage is WKT string which already does that transformation. 

Reverted in 
https://github.com/crate/crate/pull/17012/commits/974d87d2cbb73822a55020f071e76aec5be32ac9

**Comment by BaurzhanSakhariev** [comment]
> This way we will comply with the existing code in `MultiPointBuilder`/`MultiPolygonBuilder` creating `XShapeCollection` and `shape2Map` explciitly creating geometry collection for `ShapeCollection` instances.

Looks like a void concern. `shape2Map` (which could conflict with ^) is never called with an argument which went through `map2Shape`. See also https://github.com/crate/crate/pull/17012#discussion_r1848449787

**Comment by BaurzhanSakhariev** [comment]
forgot to update `test_bytes_estimate`, will push a fix

File: server/src/test/java/io/crate/types/GeoShapeTypeTest.java [hunk 2]
```diff
@@ -216,8 +216,17 @@ public void test_bytes_estimate() throws Exception {
 
     @Test
     public void test_check_equality_with_itself() throws Exception {
+        UnaryOperator<Map<String, Object>> normalizeSelect = geoJson -> type.implicitCast(geoJson);
```
Comment (seut) [inline server/src/test/java/io/crate/types/GeoShapeTypeTest.java:219]: Please always create a new unit test for a new test case

File: server/src/test/java/io/crate/types/GeoShapeTypeTest.java [hunk 3]
```diff
@@ -216,8 +216,17 @@ public void test_bytes_estimate() throws Exception {
...
+        UnaryOperator<Map<String, Object>> normalizeInsert = geoJson -> type.sanitizeValue(geoJson);
+
         for (Map<String, Object> json: geoJson) {
             assertThat(type.compare(json, json)).isEqualTo(0);
+
+            assertThat(type.compare(normalizeSelect.apply(json), json)).isEqualTo(0);
+            assertThat(type.compare(normalizeSelect.apply(json), normalizeSelect.apply(json))).isEqualTo(0);
+
+            assertThat(type.compare(normalizeInsert.apply(json), json)).isEqualTo(0);
+            assertThat(type.compare(normalizeInsert.apply(json), normalizeInsert.apply(json))).isEqualTo(0);
```
Comment (seut) [inline server/src/test/java/io/crate/types/GeoShapeTypeTest.java:229]: This will pass even without the changes. I also do not really understand how this should trip the issue. The problem is related to casting object literals containing a GeometryCollection vs a Multipoint.

File: server/src/main/java/io/crate/geo/GeoJSONUtils.java [hunk 4]
```diff
@@ -111,12 +111,13 @@ private static void forEach(Object arrayOrCollection, Consumer<Object> consumer)
...
     public static Map<String, Object> shape2Map(Shape shape) {
-        if (shape instanceof ShapeCollection) {
-            ShapeCollection<?> shapeCollection = (ShapeCollection<?>) shape;
+        if (shape instanceof ShapeCollection<?> shapeCollection) {
             List<Map<String, Object>> geometries = new ArrayList<>(shapeCollection.size());
             for (Shape collShape : shapeCollection) {
                 geometries.add(shape2Map(collShape));
             }
+            // Even though we normalize input to transform Geometry Collection to MultiPoint or MultiPolygon,
+            // We don't need to do it here as this function operates with already normalized shapes.
```
Comment (seut) [inline server/src/main/java/io/crate/geo/GeoJSONUtils.java:120]: This should better move into a method doc.  Is this related at all to the issue? Why should we normalize shapes?

File: server/src/main/java/io/crate/geo/GeoJSONUtils.java [hunk 5]
```diff
@@ -131,9 +132,14 @@ public static Map<String, Object> shape2Map(Shape shape) {
         }
     }
 
+    /**
+     * Transforms Geometry Collection to MultiPoint or MultiPolygon
+     * if all shapes in collection are points/polygons.
+     */
```
Comment (seut) [inline server/src/main/java/io/crate/geo/GeoJSONUtils.java:138]: This relates to the how `JtsSpatial` parses wtk strings. 
```suggestion
    /**
     * Convert a WKT string into a Shape.
     *
     * GeometryCollections composed of a single part are converted
     * into the respective multi-part (MultiPoint, MultiLineString, or MultiPolygon) objects
     * to maximize interoperability.
     * See {@link https://datatracker.ietf.org/doc/html/draft-ietf-geojson-03#section-3.1.8}
     */
```

File: server/src/main/java/io/crate/geo/GeoJSONUtils.java [hunk 6]
```diff
@@ -231,7 +237,14 @@ private static ShapeBuilder geoJSONString2ShapeBuilder(Map<String, Object> geoJS
...
 
-    public static void validateGeoJson(Map<?, ?> value) {
+    /**
+     *  @return validated geo json.
+     *  Transforms Geometry Collection to MultiPoint or MultiPolygon
+     *  if all shapes in collection are points/polygons.
+     *  This behavior is aligned with {@link #wkt2Shape(String)}.
+     */
+    @SuppressWarnings("unchecked")
+    public static Map<String, Object> validateGeoJson(Map<?, ?> value) {
```
Comment (seut) [inline server/src/main/java/io/crate/geo/GeoJSONUtils.java:247]:
```suggestion
    public static Map<String, Object> sanitizeMap(Map<?, ?> value) {
```

File: server/src/main/java/io/crate/geo/GeoJSONUtils.java [hunk 7]
```diff
@@ -231,7 +237,14 @@ private static ShapeBuilder geoJSONString2ShapeBuilder(Map<String, Object> geoJS
         }
     }
 
-    public static void validateGeoJson(Map<?, ?> value) {
+    /**
+     *  @return validated geo json.
+     *  Transforms Geometry Collection to MultiPoint or MultiPolygon
+     *  if all shapes in collection are points/polygons.
+     *  This behavior is aligned with {@link #wkt2Shape(String)}.
+     */
```
Comment (seut) [inline server/src/main/java/io/crate/geo/GeoJSONUtils.java:245]:
```suggestion
    /**
    * Sanitize a map of geometric values using the same constrains as {@link #wkt2Shape}
     */
```

File: server/src/main/java/io/crate/geo/GeoJSONUtils.java [hunk 8]
```diff
@@ -248,13 +261,51 @@ public static void validateGeoJson(Map<?, ?> value) {
...
-                    validateGeoJson(map);
+                    Map<String, Object> validatedShape = validateGeoJson(map);
+                    String shapeType = BytesRefs.toString(validatedShape.get(TYPE_FIELD));
+                    if (typeInCollection[0] == null) {
+                        typeInCollection[0] = shapeType;
+                    } else {
+                        if (typeInCollection[0].equals(shapeType) == false) {
+                            sameType[0] = false;
+                        }
+                    }
```
Comment (seut) [inline server/src/main/java/io/crate/geo/GeoJSONUtils.java:278]: What about nested collections? there possible converted map must be stored back as well.

File: docs/general/builtins/comparison-operators.rst [hunk 9]
```diff
@@ -80,6 +80,17 @@ This means that the order of vertices may be different in topologically equal ge
     +------+
     SELECT 1 row in set (... sec)
 
+Geometry collections, containing only points or polygons are normalized to MultiPoint and MultiPolygon.
```
Comment (seut) [inline docs/general/builtins/comparison-operators.rst:83]: Please format to max line chars of 80.
```suggestion
Geometry collections, containing only linestrings, points or polygons are normalized to MultiLineString, MultiPoint and MultiPolygon.
```

File: server/src/main/java/io/crate/geo/GeoJSONUtils.java (hunk 4 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/geo/GeoJSONUtils.java:120]: We shouldn't normalize shapes, I had a concern that we can de-normalize a valid Multi* shape and turn it to a `GeometryCollection` but I couldn't find a real scenario reflecting this.

Test with smth like `compare(..., type.implicitCast(GeoJSONUtils.map2Shape(geoJson))`) used to fail without sanitizing the result here but as it's not real scenario, I will remove this comment

File: server/src/test/java/io/crate/types/GeoShapeTypeTest.java (hunk 3 above)
Comment (BaurzhanSakhariev) [reply server/src/test/java/io/crate/types/GeoShapeTypeTest.java]: Thanks for catching, I reverted this change and added 2 tests in https://github.com/crate/crate/pull/17012/commits/ed2a9dfcd9a30c8a55d6441c34f71345c0f93ae9 (and they fail without new changes). 

One test is specifically for transformation and another to check that implicitCast/sanitizeValue (SELECT/INSERT) actually utilize new logic.

File: server/src/main/java/io/crate/geo/GeoJSONUtils.java (hunk 8 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/geo/GeoJSONUtils.java]: Thanks for catching, added handling and a test in https://github.com/crate/crate/pull/17012/commits/ed2a9dfcd9a30c8a55d6441c34f71345c0f93ae9

File: server/src/test/java/io/crate/types/GeoShapeTypeTest.java (hunk 3 above)
Comment (BaurzhanSakhariev) [reply server/src/test/java/io/crate/types/GeoShapeTypeTest.java]: I found out that I missed processing `MultiLine` string. I refactored code a bit and improved coverage in https://github.com/crate/crate/pull/17012/commits/6ad56ac442e7eda168d5352896773ed6b6277cc3

File: server/src/test/java/io/crate/types/GeoShapeTypeTest.java [hunk 10]
```diff
@@ -332,5 +332,35 @@ public void test_check_equality_of_shapes_of_same_type_with_different_coordinate
...
+        );
+        Map<String, Object> multiPoint = parse("""
+            {
+                "type": "MultiPoint",
+                "coordinates": [[1.0, 1.0], [2.0, 2.0]]
+            }
+            """
+        );
+        assertThat(type.compare(type.implicitCast(collection), type.implicitCast(multiPoint))).isEqualTo(0);
+        assertThat(type.compare(type.sanitizeValue(collection), type.sanitizeValue(multiPoint))).isEqualTo(0);
```
Comment (seut) [inline server/src/test/java/io/crate/types/GeoShapeTypeTest.java:363]: Afaik in this cases, standard map equality checks should work, no need to call `type.compare()`.
```suggestion
        assertThat(type.implicitCast(collection)).isEqualTo(type.implicitCast(multiPoint));
        assertThat(type.sanitizeValue(collection)).isEqualTo(type.sanitizeValue(multiPoint));
```

**Review by seut — APPROVED** [review]
lgtm otherwise.

**Comment by BaurzhanSakhariev** [comment]
@Mergifyio rebase

**Comment by BaurzhanSakhariev** [comment]
@Mergifyio backport 5.9
